internal/worker/sshsession/worker_test.go

```diff
@@ -0,0 +1,306 @@
+	logger sshsession.Logger,
+	modifier func(*sshsession.WorkerConfig),
+) *sshsession.WorkerConfig {
+	cg, _, _ := newStubConnectionGetter()
+	cfg := &sshsession.WorkerConfig{
+		Logger:           logger,
+		MachineId:        "1",
+		FacadeClient:     s.facadeClientMock,
+		ConnectionGetter: cg,
+	}
+
+	modifier(cfg)
+
+	return cfg
+}
+
+func (s *workerSuite) TestValidate(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	l := loggo.GetLogger("test")
+
+	// Test all OK.
+	cfg := s.newWorkerConfig(l, func(wc *sshsession.WorkerConfig) {})
+	c.Assert(cfg.Validate(), jc.ErrorIsNil)
+
+	// Test no Logger.
+	cfg = s.newWorkerConfig(
+		l,
+
+		func(cfg *sshsession.WorkerConfig) {
+			cfg.Logger = nil
+		},
+	)
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	// Test empty MachineId.
+	cfg = s.newWorkerConfig(
+		l,
+
+		func(cfg *sshsession.WorkerConfig) {
+			cfg.MachineId = ""
+		},
+	)
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	// Test no FacadeClient.
+	cfg = s.newWorkerConfig(
+		l,
+
+		func(cfg *sshsession.WorkerConfig) {
+			cfg.FacadeClient = nil
+		},
+	)
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+
+	// Test no ConnectionGetter.
+	cfg = s.newWorkerConfig(
+		l,
+
+		func(cfg *sshsession.WorkerConfig) {
+			cfg.ConnectionGetter = nil
+		},
+	)
+	c.Assert(cfg.Validate(), jc.ErrorIs, errors.NotValid)
+}
+
+func (s *workerSuite) TestSSHSessionWorkerCanBeKilled(c *gc.C) {
+	ctrl := s.setupMocks(c)
+	defer ctrl.Finish()
+
+	l := loggo.GetLogger("test")
+
+	stringChan := watcher.StringsChannel(make(chan []string))
+	s.watcherMock.EXPECT().Changes().Return(stringChan).AnyTimes()
+	s.facadeClientMock.EXPECT().WatchSSHConnRequest("0").Return(s.watcherMock, nil).AnyTimes()
+
+	// Check the water is Wait()'ed and Kill()'ed exactly once.
+	s.watcherMock.EXPECT().Wait().Times(1)
+	s.watcherMock.EXPECT().Kill().Times(1)
+
+	connGetter, _, _ := newStubConnectionGetter()
+	defer connGetter.close()
+
+	w, err := sshsession.NewWorker(sshsession.WorkerConfig{
+		Logger:           l,
+		MachineId:        "0",
+		FacadeClient:     s.facadeClientMock,
+		ConnectionGetter: connGetter,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(workertest.CheckKill(c, w), jc.ErrorIsNil)
+}
+
+// TestSSHSessionWorkerHandlesConnection tests that the worker can at least pipe the
+// connections together using an in-memory net.Pipe. Other than an actual integration
+// test, we cannot test the literal SSH connections to the controller and local SSHD.
+//
+// Additionally, as a side effect, we're testing the ephemeral key is added and deleted.
+// We don't check explicitly for jujussh.AddKeys as it will error on failure anyway.
+// So we only explicitly check that the key has been removed from the authorized_key file
+// on connection closure. We could write whitebox tests for key addition/removal, but we'd
+// really just be testing the package.
+func (s *workerSuite) TestSSHSessionWorkerHandlesConnection(c *gc.C) {
```

Comment:
We need to look closely at our use of sleeps in this test and address them because it seems unlikely to me that we need them at all.

Reply:
Fixed.